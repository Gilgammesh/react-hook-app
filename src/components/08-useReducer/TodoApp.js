import React, { useEffect, useReducer } from "react";
import { todoReducer } from "./todoReducer";
import TodoForm from "./TodoForm";
import Todo from "./Todo";
import "./styles.css";

// const initialState = [
//   {
//     id: new Date().getTime(),
//     desc: "Aprender React",
//     done: false,
//   },
// ];

const init = () => {
  return JSON.parse(localStorage.getItem("todos")) || [];
};

const TodoApp = () => {
  //   const [todos, dispatch] = useReducer(todoReducer, initialState, init);
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos));
  }, [todos]);

  const todosList = todos.map((ele, index) => {
    return <Todo key={ele.id} ele={ele} index={index} dispatch={dispatch} />;
  });

  return (
    <div>
      <h1>Todo App ({todos.length})</h1>
      <hr />
      <TodoForm dispatch={dispatch} />
      <hr />
      <h3>Tareas</h3>
      <div className="row">{todosList}</div>
    </div>
  );
};

export default TodoApp;
