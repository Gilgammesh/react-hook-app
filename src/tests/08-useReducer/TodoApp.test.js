import React from "react";
import { shallow, mount } from "enzyme";
import { act } from "@testing-library/react";
import "@testing-library/jest-dom";
import TodoApp from "../../components/08-useReducer/TodoApp";
import { demoTodos } from "../fixtures/demoTodos";

describe("Pruebas unitarias en el componente TodoApp.js", () => {
  Storage.prototype.setItem = jest.fn(() => {});
  /************************************************************************************/
  test("se debe mostrar el componente <TodoApp /> correctamente", () => {
    const wrapper = shallow(<TodoApp />);
    expect(wrapper).toMatchSnapshot();
  });
  /************************************************************************************/
  test("se debe agregar un TODO y cambiar el localstorage", () => {
    const value = "Aprender React";
    const wrapper = mount(<TodoApp />);
    act(() => {
      const dispatch = wrapper.find("TodoForm").prop("dispatch");
      dispatch({
        payload: demoTodos[0],
        type: "add",
      });
      dispatch({
        payload: demoTodos[1],
        type: "add",
      });
    });
    const h1 = wrapper.find('h1');
    expect(h1.text().trim()).toBe(`Todo App (2)`);
    expect(localStorage.setItem).toHaveBeenCalledTimes(2);    
  });
  /************************************************************************************/
  test("se debe eliminar un TODO y cambiar el localstorage", () => {
    const value = "Aprender React";
    const wrapper = mount(<TodoApp />);
    act(() => {
      const dispatch = wrapper.find("TodoForm").prop("dispatch");
      dispatch({
        payload: demoTodos[0],
        type: "add",
      });
      dispatch({
        payload: demoTodos[0].id,
        type: "delete",
      });
    });
    const h1 = wrapper.find('h1');
    expect(h1.text().trim()).toBe(`Todo App (0)`);
    expect(localStorage.setItem).toHaveBeenCalledTimes(2);    
  });
});
