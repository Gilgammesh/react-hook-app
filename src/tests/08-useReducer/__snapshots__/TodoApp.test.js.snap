// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas unitarias en el componente TodoApp.js se debe mostrar el componente <TodoApp /> correctamente 1`] = `
<div>
  <h1>
    Todo App (
    0
    )
  </h1>
  <hr />
  <TodoForm
    dispatch={[Function]}
  />
  <hr />
  <h3>
    Tareas
  </h3>
  <div
    className="row"
  />
</div>
`;
